packages/shopify-app-remix/src/server/authenticate/public/appProxy/__tests__/authenticate.test.ts

```diff
@@ -0,0 +1,280 @@
+import {HashFormat, createSHA256HMAC} from '@shopify/shopify-api/runtime';
+
+import {LATEST_API_VERSION, shopifyApp} from '../../../..';
+import {
+  API_SECRET_KEY,
+  APP_URL,
+  TEST_SHOP,
+  getThrownResponse,
+  setUpValidSession,
+  testConfig,
+} from '../../../../__tests__/test-helper';
+import {mockExternalRequest} from '../../../../__tests__/request-mock';
+
+const REQUEST_URL = `https://${TEST_SHOP}/admin/api/${LATEST_API_VERSION}/customers.json`;
+
+describe('authenticating app proxy requests', () => {
+  it('Throws a 400 response if there is no signature param', async () => {
+    // GIVEN
+    const shopify = shopifyApp(testConfig());
+
+    // WHEN
+    const url = new URL(APP_URL);
+    url.searchParams.set('shop', TEST_SHOP);
+    url.searchParams.set('timestamp', secondsInPast(10));
+
+    const response = await getThrownResponse(
+      shopify.authenticate.public.appProxy,
+      new Request(url.toString()),
+    );
+
+    // THEN
+    expect(response.status).toBe(400);
+    expect(response.statusText).toBe('Bad Request');
+  });
+
+  it('Throws a 400 response if the signature param is incorrect', async () => {
+    // GIVEN
+    const shopify = shopifyApp(testConfig());
+
+    // WHEN
+    const url = new URL(APP_URL);
+    url.searchParams.set('shop', TEST_SHOP);
+    url.searchParams.set('timestamp', secondsInPast(10));
+    url.searchParams.set('signature', await createAppProxyHmac(url));
+
+    const response = await getThrownResponse(
+      shopify.authenticate.public.appProxy,
+      new Request(url.toString()),
+    );
+
+    // THEN
+    expect(response.status).toBe(400);
+    expect(response.statusText).toBe('Bad Request');
+  });
+
+  it('Throws a 400 response if the timestamp is more than 90 seconds old', async () => {
+    // GIVEN
+    const shopify = shopifyApp(testConfig());
+
+    // WHEN
+    const url = new URL(APP_URL);
+    url.searchParams.set('shop', TEST_SHOP);
+    url.searchParams.set('timestamp', secondsInPast(100));
+    url.searchParams.set('signature', await createAppProxyHmac(url));
+
+    const response = await getThrownResponse(
+      shopify.authenticate.public.appProxy,
+      new Request(url.toString()),
+    );
+
+    // THEN
+    expect(response.status).toBe(400);
+    expect(response.statusText).toBe('Bad Request');
+  });
+
+  it('Throws a 400 response if the timestamp is more than 90 seconds in the future', async () => {
```

Comment:
Is it worth testing that <90s in the future still works?

Reply:
Yeah, for sure.  Funny I didn't include a test for the success state.  Added.